src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -875,27 +885,61 @@ private bool IsAttributeDeclarationTerminator()
 
-        private AttributeListSyntax ParseAttributeDeclaration()
+        private AttributeListSyntax? TryParseAttributeDeclaration(bool inExpressionContext)
         {
-            if (this.IsIncrementalAndFactoryContextMatches && this.CurrentNodeKind == SyntaxKind.AttributeList)
+            if (this.IsIncrementalAndFactoryContextMatches &&
+                this.CurrentNodeKind == SyntaxKind.AttributeList &&
+                !inExpressionContext)
             {
                 return (AttributeListSyntax)this.EatNode();
             }
 
+            using var resetPoint = GetDisposableResetPoint(resetOnDispose: false);
+
             var openBracket = this.EatToken(SyntaxKind.OpenBracketToken);
 
             // Check for optional location :
             var location = IsSomeWord(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken
                 ? _syntaxFactory.AttributeTargetSpecifier(ConvertToKeyword(this.EatToken()), this.EatToken(SyntaxKind.ColonToken))
                 : null;
 
-            var attributes = _pool.AllocateSeparated<AttributeSyntax>();
-            this.ParseAttributes(attributes);
-            return _syntaxFactory.AttributeList(
-                openBracket,
-                location,
-                _pool.ToListAndFree(attributes),
-                this.EatToken(SyntaxKind.CloseBracketToken));
+            var attributesBuilder = _pool.AllocateSeparated<AttributeSyntax>();
+            this.ParseAttributes(attributesBuilder);
+            var attributes = _pool.ToListAndFree(attributesBuilder);
+
+            var closeBracket = this.EatToken(SyntaxKind.CloseBracketToken);
+            if (inExpressionContext && ParseAsCollectionLiteral())
+            {
+                // we're in an expression and we've seen `[A, B].`  This is actually the start of a collection literal
+                // that someone is explicitly accessing a member off of.
+                resetPoint.Reset();
+                return null;
+            }
+
+            return _syntaxFactory.AttributeList(openBracket, location, attributes, closeBracket);
+
+            bool ParseAsCollectionLiteral()
+            {
+                // `[A, B].` is a member access off of a collection literal. 
+                if (this.CurrentToken.Kind == SyntaxKind.DotToken)
+                    return true;
+
+                // `[A, B]->` is a member access off of a collection literal. Note: this will always be illegal
+                // semantically (as a collection literal has the natural type List<> which is not a pointer type).  But
+                // we leave that check to binding.
+                if (this.CurrentToken.Kind == SyntaxKind.MinusGreaterThanToken)
+                    return true;
+
+                // `[A, B]?.`  The `?` is unnecessary (as a collection literal is always non-null).  But it is fine if
+                // the use wants to include it unnecessarily.  A linter could always advise they remove it.
```

Comment:
Consider removing.

Reply:
Which part?  The entire comment, or the linter part?  Thanks!

Reply:
I meant consider removing the last two or three sentences.

Reply:
ok.